### PR TITLE
Laue diffuse

### PR DESCRIPTION
**Pull request by dermen**
Mike's and David's PR for better diffuse scattering models in diffBragg

### CODE REVIEW

**Comment by dermen** [comment]
one thing I noticed: if you run 

```
DIFFBRAGG_USE_CUDA=1 libtbx.python tst_diffBragg_diffuse_properties --gamma 100 150 200 --laue
```

the finite differences errors do not approach 0, rather they hover around 1 . Similar behavior for `--stencil 1`

Investigating ...

**Comment by dermen** [comment]
latest commit fixes aforementioned test issue .. the test is not run automatically ...

**Comment by dermen** [comment]
From my end this looks good to merge in ... unless we want to quickly add the CPU support

**Comment by dermen** [comment]
@mewall - I think maybe you didnt commit the new diffuse_util.h file ? I really like the new method, it will likely help any future kokkos work - I should adopt this for the derivative terms as well ..

**Comment by mewall** [comment]
Just force-pushed an amended commit including that file. Yes, perhaps it can help with the kokkos development...

________________________________
From: Derek Mendez ***@***.***>
Sent: Tuesday, August 30, 2022 7:28:47 PM
To: cctbx/cctbx_project
Cc: Wall, Michael E; Mention
Subject: [EXTERNAL] Re: [cctbx/cctbx_project] Laue diffuse (PR #785)


@mewall<https://urldefense.com/v3/__https://github.com/mewall__;!!Bt8fGhp8LhKGRg!EO6kEHOuXj5mOzoOfU2WiFykMHJIP7RoSG-FcxY--B32BjKLkeNe6mLVtS2dJYhxHynz696MsqpAGBml7mFJmsAI$> - I think maybe you didnt commit the new diffuse_util.h file ? I really like the new method, it will likely help any future kokkos work - I should adopt this for the derivative terms as well ..

File: simtbx/diffBragg/phil.py [hunk 1]
```diff
@@ -542,6 +542,9 @@
   .help = if True, let the values of init.diffuse_sigma and init.diffuse_gamma
   .help = be used to define the diffuse scattering. Set e.g. fix.diffuse_sigma=True in order to refine them
   .expert_level = 10
+symmetrize_diffuse = False
```
Comment (mewall) [inline simtbx/diffBragg/phil.py:545]: @dermen How about we make the default here true?

File: simtbx/diffBragg/tests/tst_diffBragg_diffuse_properties.py [hunk 2]
```diff
@@ -11,15 +11,25 @@
...
 
 import pylab as plt
+import os
+
 args = parser.parse_args()
 if args.cuda:
-    import os
     os.environ["DIFFBRAGG_USE_CUDA"] = "1"
 
+if os.environ.get("DIFFBRAGG_USE_CUDA") is None:
```
Comment (mewall) [inline simtbx/diffBragg/tests/tst_diffBragg_diffuse_properties.py:25]: @dermen We can remove this test now, correct?

File: simtbx/diffBragg/phil.py (hunk 1 above)
Comment (dermen) [reply simtbx/diffBragg/phil.py:545]: I guess it has potential to increase the iteration time (maybe not as much as stenciling), which is why I left it False by default .. Happy to toggle it though

File: simtbx/diffBragg/tests/tst_diffBragg_diffuse_properties.py (hunk 2 above)
Comment (dermen) [reply simtbx/diffBragg/tests/tst_diffBragg_diffuse_properties.py:25]: I believe so, though I havent tested myself

**Comment by mewall** [comment]
Let's set this one to True by default as I believe the model is actually incorrect without the symmetrization...


But it's OK to leave the stencil size at 0.

________________________________
From: Derek Mendez ***@***.***>
Sent: Wednesday, August 31, 2022 5:59:16 PM
To: cctbx/cctbx_project
Cc: Wall, Michael E; Mention
Subject: [EXTERNAL] Re: [cctbx/cctbx_project] Laue diffuse (PR #785)


@dermen commented on this pull request.

________________________________

In simtbx/diffBragg/phil.py<https://urldefense.com/v3/__https://github.com/cctbx/cctbx_project/pull/785*discussion_r960113065__;Iw!!Bt8fGhp8LhKGRg!CNGXzKP8ZvC7jOAhqyR-kz36hyuJobBT899NuxMOqiXSAUfbpfDpgRdnitn_FIDzMm-HCeYwsbLJ_iyEYw6ndddo$>:

> @@ -542,6 +542,9 @@
   .help = if True, let the values of init.diffuse_sigma and init.diffuse_gamma
   .help = be used to define the diffuse scattering. Set e.g. fix.diffuse_sigma=True in order to refine them
   .expert_level = 10
+symmetrize_diffuse = False


I guess it has potential to increase the iteration time (maybe not as much as stenciling), which is why I left it False by default .. Happy to toggle it though

File: simtbx/diffBragg/tests/tst_diffBragg_diffuse_properties.py (hunk 2 above)
Comment (mewall) [reply simtbx/diffBragg/tests/tst_diffBragg_diffuse_properties.py:25]: OK. I think the CPU code should be ready to use now, so let's just use it here...

I can do the edits tomorrow morning (for me) if you don't get around to it.

File: simtbx/diffBragg/src/diffuse_util.h [hunk 3]
```diff
@@ -0,0 +1,639 @@
+#ifndef SIMTBX_DIFFBRAGG_DIFFUSE_UTIL
+#define SIMTBX_DIFFBRAGG_DIFFUSE_UTIL
+
+#include <simtbx/diffBragg/src/diffuse_util.h>
```
Comment (Trzs) [inline simtbx/diffBragg/src/diffuse_util.h:4]: The file currently includes itself, did you mean to include utils.h?

Comment (mewall) [reply simtbx/diffBragg/src/diffuse_util.h]: This has been addressed. thanks

File: simtbx/diffBragg/tests/tst_diffBragg_diffuse_properties.py (hunk 2 above)
Comment (mewall) [reply simtbx/diffBragg/tests/tst_diffBragg_diffuse_properties.py]: Updated.

File: simtbx/diffBragg/phil.py (hunk 1 above)
Comment (mewall) [reply simtbx/diffBragg/phil.py]: Made default Tru in subsequent commit

**Review by mewall — APPROVED** [review]
@bkpoon @dermen @nksauter Billy: As a heads up, tests are failing due to a brown-out in availability of Mac OS CI images. Let me know if there are any concerns, otherwise I will plan to merge into master after discussing this PR with Nick today, assuming other CI tests pass.

**Comment by nksauter** [comment]
For completeness, note that the symmetry operators can be generated programmatically with this example script:
```
from __future__ import absolute_import, division, print_function
from cctbx import sgtbx
from cctbx.sgtbx import space_group_symbols, space_group_info, space_group_type
import re

def get_laue_group_number(sg_symbol=None):
    """ get laue group number from space group symbol """
    if sg_symbol is None:
        laue_sym = "P-1"
    else:
        g = sgtbx.space_group_info(sg_symbol).group()
        if g.laue_group_type() not in ["2/m", "-3m"]:
            laue_sym = "P{}".format(g.laue_group_type())
        else:
            pg = str(g.build_derived_patterson_group().info().symbol_and_number())
            lc = re.sub(r'\([^)]*\)', '', pg[2:]).replace(" ", "")
            if pg[0] == "R":
                lc = lc.replace(":H", "1")
            laue_sym = "P{}".format(lc)

    hm_symbols = ['P-1', 'P112/m', 'P12/m1', 'P2/m11', 'Pmmm', 'P4/m', 'P4/mmm', 'P-3', 'P-3m1', 'P-31m', 'P6/m',
                  'P6/mmm', 'Pm-3', 'Pm-3m']
    lgs = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
    return lgs[hm_symbols.index(laue_sym)], laue_sym

if __name__=="__main__":
  for sgnum in range(1,231):
    SYM = space_group_symbols(sgnum)
    SG = sgtbx.space_group(SYM)
    SGT = SG.type()
    lgs, laue_sym = get_laue_group_number(sg_symbol=SYM.hermann_mauguin())
    
    print (sgnum, "%15s"%(SGT.lookup_symbol()),  "%10s"%laue_sym)
    laue_sym_as_group = space_group_info(laue_sym).group()

    for i in laue_sym_as_group.smx():
      print ("%15s"%i,  ", ".join( ["%2d"%(i.r().as_rational()[arg]) for arg in range(9)] )  )
    print ("   all",laue_sym_as_group.n_smx())
```

**Comment by mewall** [comment]
Fixed, thanks.

________________________________
From: Felix Wittwer ***@***.***>
Sent: Thursday, September 1, 2022 11:55:48 AM
To: cctbx/cctbx_project
Cc: Wall, Michael E; Mention
Subject: [EXTERNAL] Re: [cctbx/cctbx_project] Laue diffuse (PR #785)


@Trzs commented on this pull request.

________________________________

In simtbx/diffBragg/src/diffuse_util.h<https://urldefense.com/v3/__https://github.com/cctbx/cctbx_project/pull/785*discussion_r960953564__;Iw!!Bt8fGhp8LhKGRg!Df0S5df56n7HoMxIs8yH4VhxMFRb-54tW5U-qUGtSdWG1ek2jQuHoNE_5PqVKlLfhj4vb2wLVMqa_WucfaF-Pp6O$>:

> @@ -0,0 +1,639 @@
+#ifndef SIMTBX_DIFFBRAGG_DIFFUSE_UTIL
+#define SIMTBX_DIFFBRAGG_DIFFUSE_UTIL
+
+#include <simtbx/diffBragg/src/diffuse_util.h>


The file currently includes itself, did you mean to include utils.h?

**Pull request merged**